{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f4d6196",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gzip\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1089c982",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the full GWAS catalog\n",
    "df = pd.read_csv(\"./data/vcf/gwas-catalog-associations.tsv\", sep='\\t', low_memory=False) # 2018-12-27\n",
    "\n",
    "# Filter for disease\n",
    "disease = 'type 2 diabetes'\n",
    "disease_file = disease.replace(\" \", \"_\")\n",
    "df_disease = df[df['DISEASE/TRAIT'].str.contains(disease, case=False, na=False)]\n",
    "\n",
    "# Drop rows missing position info\n",
    "df_disease = df_disease.dropna(subset=['CHR_ID', 'CHR_POS'])\n",
    "\n",
    "# Convert to VCF-style ids and positions\n",
    "df_disease['CHR_ID'] = df_disease['CHR_ID'].astype(str).str.split(';').str[0] # Keep only the first value before any semicolon\n",
    "df_disease['CHR_POS'] = df_disease['CHR_POS'].astype(str).str.split(';').str[0] # Keep only the first value before any semicolon\n",
    "df_disease = df_disease[df_disease['CHR_POS'].astype(str).str.match(r'^\\d+$')]\n",
    "df_disease['CHR_POS'] = df_disease['CHR_POS'].astype(int)\n",
    "df_disease = df_disease[df_disease['DISEASE/TRAIT']=='Type 2 diabetes'] # Keep only the rows that are just type 2 diabetes\n",
    "\n",
    "# Create a directory only if it doesn't exist\n",
    "folder_path = f\"./data/vcf/{disease_file}\"\n",
    "if not os.path.exists(folder_path):\n",
    "    os.mkdir(folder_path)\n",
    "    \n",
    "# Save to tsv\n",
    "df_disease[['CHR_ID', 'CHR_POS']].to_csv(f\"./data/vcf/{disease_file}/{disease_file}_variants.tsv\", sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ae70190",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter each chromosome VCF file for variants\n",
    "for i in range(1, 23):\n",
    "    vcf_filter_command = [\n",
    "        \"bcftools\", \"view\",\n",
    "        \"-R\", f\"./data/vcf/{disease_file}/{disease_file}_variants.tsv\",\n",
    "        f\"./data/vcf/ALL.chr{i}.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz\",\n",
    "        \"-Oz\", \"-o\", f\"./data/vcf/{disease_file}/{disease_file}_chr{i}.vcf.gz\"\n",
    "    ]\n",
    "\n",
    "    subprocess.run(vcf_filter_command, check=True)\n",
    "\n",
    "# Files to combine\n",
    "vcf_file = f'ls ./data/vcf/{disease_file}/{disease_file}_chr*.vcf.gz > ./data/vcf/{disease_file}/vcf_list.txt'\n",
    "subprocess.run(vcf_file, shell=True, check=True)\n",
    "\n",
    "# Combine all filtered files into one VCF file\n",
    "vcf_concat_command = [\n",
    "    \"bcftools\", \"concat\",\n",
    "    \"-f\", f\"./data/vcf/{disease_file}/vcf_list.txt\",\n",
    "    \"-Oz\", \"-o\", f\"./data/vcf/{disease_file}/{disease_file}_all.vcf.gz\"\n",
    "]\n",
    "subprocess.run(vcf_concat_command, check=True)\n",
    "\n",
    "# Filter VCF for substitutions only\n",
    "sub_filter = [\n",
    "    \"bcftools\", \"view\",\n",
    "    \"-e\", 'ALT ~ \"<\"',\n",
    "    f\"./data/vcf/{disease_file}/{disease_file}_all.vcf.gz\",\n",
    "    \"-Oz\",\n",
    "    \"-o\", f\"./data/vcf/{disease_file}/{disease_file}_all_substitution.vcf.gz\"\n",
    "]\n",
    "subprocess.run(sub_filter, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5291c245",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read vcf as DataFrame for sanity check\n",
    "\n",
    "# Filepath\n",
    "vcf_path = f\"./data/vcf/{disease_file}/{disease_file}_all_substitution.vcf.gz\"\n",
    "\n",
    "# Find the header line\n",
    "with gzip.open(vcf_path, 'rt') as f:\n",
    "    for i, line in enumerate(f):\n",
    "        if line.startswith(\"#CHROM\"):\n",
    "            header_line_num = i\n",
    "            break\n",
    "\n",
    "# Read the file as DataFrame\n",
    "df_disease_vcf = pd.read_csv(\n",
    "    vcf_path,\n",
    "    sep='\\t',\n",
    "    compression='gzip',\n",
    "    skiprows=header_line_num,\n",
    "    header=0\n",
    ")\n",
    "\n",
    "df_disease_vcf"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
